stdlib/public/core/CompilerProtocols.swift

```diff
@@ -178,37 +178,42 @@ public func != <T : Equatable>(lhs: T, rhs: T) -> Bool
   return lhs.rawValue != rhs.rawValue
 }
 
-/// A type that can produce a collection of all of its values.
+/// A type that provides a collection of all of its values.
 ///
-/// Simple Enumerations
-/// ===================
+/// Types that conform to the `CaseIterable` protocol are typically
+/// enumerations without associated values. When using a `CaseIterable` type,
+/// you can access a collection of all of the type's cases by using the type's
+/// `allCases` property.
 ///
-/// For any Swift enumeration where every case does not have an associated
-/// value, the Swift compiler can automatically fill out the `CaseIterable`
-/// conformance. When defining your own custom enumeration, specify a
-/// conformance to `CaseIterable` to take advantage of this automatic
-/// derivation.
+/// For example, the `CompassDirection` enumeration declared in this example
+/// conforms to `CaseIterable`. You access the number of cases and the cases
+/// themselves through `CompassDirection.allCases`.
 ///
-/// For example, every case of the `CardinalPoint` enumeration defined here
-/// does not have an associated value:
-///
-///     enum CardinalPoint: CaseIterable {
+///     enum CompassDirection: CaseIterable {
 ///         case north, south, east, west
 ///     }
 ///
-/// So the compiler automatically creates a conformance.
+///     print("There are \(CompassDirection.allCases.count) directions.")
+///     // Prints "There are 4 directions."
+///     let caseList = CompassDirection.allCases
+///                                    .map({ "\($0)" })
+///                                    .joined(separator: ", "))
+///     // caseList == "north, south, east, west"
 ///
-///     for cardinality in CardinalPoint.allCases {
-///         print(cardinality)
-///     }
-///     // Prints "north"
-///     // Prints "south"
-///     // Prints "east"
-///     // Prints "west"
+/// Conforming to the CaseIterable Protocol
+/// =======================================
+///
+/// The compiler can automatically provide an implementation of the
+/// `CaseIterable` requirements for any enumeration without associated values.
+/// To take advantage of this automatic synthesis when defining your own custom
+/// enumeration, specify conformance to the `CaseIterable` protocol in the
+/// original declaration.
```

Comment:
There is a caveat to this that may or may not be worth documenting: Availability attributes on any cases suppress synthesis.

Reply:
This is a limitation of the current implementation only.

Reply:
Thanks, I do want to make sure that's covered.

Reply:
> Cases marked unavailable aren't included in the synthesized allCases

This should always be the case - we do not want the protocol to allow users to break the availability model, even when/if we generalize to remove the restriction

> Any @available attribute suppresses the synthesis for the entire enumeration

This is the reality in the field, definitely not an ideal.

Reply:
I've documented the reality. When the rule is relaxed, we can switch the `@available` limitation with a note about `unavailable` cases.

Reply:
Thanks for the feedback! 🙇